# Setup

## Loading libraries

In [4]:
import requests # To get data
import pandas as pd

## Setting global variables
According to the documentation, the URL to retrieve the data from the World Bank API is the following:

In [5]:
WB_URL = 'http://api.worldbank.org/v2/country/all/indicator/SH.STA.BASS.ZS'

# Getting data from the API

## Retrieving data

### 1st try: data returned in XML format

In [6]:
response = requests.get(WB_URL)
response.text

'<?xml version="1.0" encoding="utf-8"?>\r\n<wb:data page="1" pages="330" per_page="50" total="16492" sourceid="2" sourcename="World Development Indicators" lastupdated="2023-03-30" xmlns:wb="http://www.worldbank.org">\r\n  <wb:data>\r\n    <wb:indicator id="SH.STA.BASS.ZS">People using at least basic sanitation services (% of population)</wb:indicator>\r\n    <wb:country id="ZH">Africa Eastern and Southern</wb:country>\r\n    <wb:countryiso3code>AFE</wb:countryiso3code>\r\n    <wb:date>2021</wb:date>\r\n    <wb:value />\r\n    <wb:unit />\r\n    <wb:obs_status />\r\n    <wb:decimal>0</wb:decimal>\r\n  </wb:data>\r\n  <wb:data>\r\n    <wb:indicator id="SH.STA.BASS.ZS">People using at least basic sanitation services (% of population)</wb:indicator>\r\n    <wb:country id="ZH">Africa Eastern and Southern</wb:country>\r\n    <wb:countryiso3code>AFE</wb:countryiso3code>\r\n    <wb:date>2020</wb:date>\r\n    <wb:value>30.8359427084217</wb:value>\r\n    <wb:unit />\r\n    <wb:obs_status />\

### 2nd try: data returned in JSON format
According to the [documentation](https://datahelpdesk.worldbank.org/knowledgebase/articles/898581-api-basic-call-structures), the data can be returned in JSON format. 

In [7]:
response = requests.get(WB_URL+'?format=json')
response.text

'[{"page":1,"pages":330,"per_page":50,"total":16492,"sourceid":"2","sourcename":"World Development Indicators","lastupdated":"2023-03-30"},[{"indicator":{"id":"SH.STA.BASS.ZS","value":"People using at least basic sanitation services (% of population)"},"country":{"id":"ZH","value":"Africa Eastern and Southern"},"countryiso3code":"AFE","date":"2021","value":null,"unit":"","obs_status":"","decimal":0},{"indicator":{"id":"SH.STA.BASS.ZS","value":"People using at least basic sanitation services (% of population)"},"country":{"id":"ZH","value":"Africa Eastern and Southern"},"countryiso3code":"AFE","date":"2020","value":30.8359427084217,"unit":"","obs_status":"","decimal":0},{"indicator":{"id":"SH.STA.BASS.ZS","value":"People using at least basic sanitation services (% of population)"},"country":{"id":"ZH","value":"Africa Eastern and Southern"},"countryiso3code":"AFE","date":"2019","value":30.6558737186025,"unit":"","obs_status":"","decimal":0},{"indicator":{"id":"SH.STA.BASS.ZS","value":"Pe

In [8]:
# Get json content
data = response.json()
data

[{'page': 1,
  'pages': 330,
  'per_page': 50,
  'total': 16492,
  'sourceid': '2',
  'sourcename': 'World Development Indicators',
  'lastupdated': '2023-03-30'},
 [{'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
   'countryiso3code': 'AFE',
   'date': '2021',
   'value': None,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
   'countryiso3code': 'AFE',
   'date': '2020',
   'value': 30.8359427084217,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
   'coun

### 3rd try: getting data from 1960
According to the [documentation](https://datahelpdesk.worldbank.org/knowledgebase/articles/898581-api-basic-call-structures), requests support year-to-date values (YTD) and we are interested in pulling data from 1960

In [9]:
response = requests.get(WB_URL+'?format=json&YTD:1960')
# Get json content
data = response.json()
data

[{'page': 1,
  'pages': 330,
  'per_page': 50,
  'total': 16492,
  'sourceid': '2',
  'sourcename': 'World Development Indicators',
  'lastupdated': '2023-03-30'},
 [{'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
   'countryiso3code': 'AFE',
   'date': '2021',
   'value': None,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
   'countryiso3code': 'AFE',
   'date': '2020',
   'value': 30.8359427084217,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'},
   'coun

### 4th try: getting data from different pages
According to API's response, there are 330 pages. We are interested in retrieving all of them.

In [10]:
response = requests.get(WB_URL+'?format=json&YTD:1960&page=200')
# Get json content
data = response.json()
data

[{'page': 200,
  'pages': 330,
  'per_page': 50,
  'total': 16492,
  'sourceid': '2',
  'sourcename': 'World Development Indicators',
  'lastupdated': '2023-03-30'},
 [{'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'LS', 'value': 'Lesotho'},
   'countryiso3code': 'LSO',
   'date': '1991',
   'value': None,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'LS', 'value': 'Lesotho'},
   'countryiso3code': 'LSO',
   'date': '1990',
   'value': None,
   'unit': '',
   'obs_status': '',
   'decimal': 0},
  {'indicator': {'id': 'SH.STA.BASS.ZS',
    'value': 'People using at least basic sanitation services (% of population)'},
   'country': {'id': 'LS', 'value': 'Lesotho'},
   'countryiso3code': 'LSO',
   'date': '1989',
   'value': None,
   'unit': '

## Parsing data into a dataframe

### 1st try: Using DataFrame
The data is not parsed as expected

In [11]:
data_df = pd.DataFrame(data[1])
data_df.head()

indicator   
0  {'id': 'SH.STA.BASS.ZS', 'value': 'People usin...  \
1  {'id': 'SH.STA.BASS.ZS', 'value': 'People usin...   
2  {'id': 'SH.STA.BASS.ZS', 'value': 'People usin...   
3  {'id': 'SH.STA.BASS.ZS', 'value': 'People usin...   
4  {'id': 'SH.STA.BASS.ZS', 'value': 'People usin...   

                            country countryiso3code  date  value unit   
0  {'id': 'LS', 'value': 'Lesotho'}             LSO  1991    NaN       \
1  {'id': 'LS', 'value': 'Lesotho'}             LSO  1990    NaN        
2  {'id': 'LS', 'value': 'Lesotho'}             LSO  1989    NaN        
3  {'id': 'LS', 'value': 'Lesotho'}             LSO  1988    NaN        
4  {'id': 'LS', 'value': 'Lesotho'}             LSO  1987    NaN        

  obs_status  decimal  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0

### 2nd try: Using json_normalize
This option is good enough

In [12]:
data_df = pd.json_normalize(data[1])
data_df.head()

countryiso3code  date  value unit obs_status  decimal    indicator.id   
0             LSO  1991    NaN                        0  SH.STA.BASS.ZS  \
1             LSO  1990    NaN                        0  SH.STA.BASS.ZS   
2             LSO  1989    NaN                        0  SH.STA.BASS.ZS   
3             LSO  1988    NaN                        0  SH.STA.BASS.ZS   
4             LSO  1987    NaN                        0  SH.STA.BASS.ZS   

                                     indicator.value country.id country.value  
0  People using at least basic sanitation service...         LS       Lesotho  
1  People using at least basic sanitation service...         LS       Lesotho  
2  People using at least basic sanitation service...         LS       Lesotho  
3  People using at least basic sanitation service...         LS       Lesotho  
4  People using at least basic sanitation service...         LS       Lesotho

In [13]:
print(data_df['countryiso3code'].unique())
print(data_df['date'].unique())
print(data_df['value'].unique())
print(data_df['unit'].unique())
print(data_df['obs_status'].unique())
print(data_df['decimal'].unique())
print(data_df['indicator.id'].unique())
print(data_df['indicator.value'].unique())
print(data_df['country.id'].unique())
print(data_df['country.value'].unique())

['LSO' 'LBR']
['1991' '1990' '1989' '1988' '1987' '1986' '1985' '1984' '1983' '1982'
 '1981' '1980' '1979' '1978' '1977' '1976' '1975' '1974' '1973' '1972'
 '1971' '1970' '1969' '1968' '1967' '1966' '1965' '1964' '1963' '1962'
 '1961' '1960' '2021' '2020' '2019' '2018' '2017' '2016' '2015' '2014'
 '2013' '2012' '2011' '2010' '2009' '2008' '2007' '2006' '2005' '2004']
[        nan 18.16357204 18.05615325 17.6760717  17.49178368 17.30317359
 17.11000494 16.9129365  16.71216861 16.50766421 16.30006099 16.08907782
 15.87533316 15.65875523 15.43948847 15.21515877 14.98613608 14.61709797]
['']
['']
[0]
['SH.STA.BASS.ZS']
['People using at least basic sanitation services (% of population)']
['LS' 'LR']
['Lesotho' 'Liberia']


## Pulling all data from the API
Now, that we know how to get the data from the API (4th version in **Retrieving section**) and parse it into a dataframe (2nd version in **Parsing data**), we can pull all the data.

In [31]:
# Getting the number of pages
temp_response = requests.get(WB_URL+'?format=json&YTD:1960')
temp_response = temp_response.json()
# temp_response
# temp_response[0]
# temp_response[0]['pages']
n_pages = temp_response[0]['pages']
total_entries = temp_response[0]['total']
print('Total pages: ' + str(n_pages))
print('Total entries: ' + str(total_entries))

Total pages: 330
Total entries: 16492


In [27]:
# Loop to retrieve all the data
sanitation_df = pd.DataFrame()
for i in range(n_pages):
    print('Page {} / {}'.format(i,n_pages))
    temp_response = requests.get(WB_URL+'?format=json&YTD:1960&page='+str(i+1))
    temp_data = temp_response.json()
    temp_df = pd.json_normalize(temp_data[1])
    temp_df['page'] = i
    sanitation_df = pd.concat([sanitation_df, temp_df])

Page 0 / 330
Page 1 / 330
Page 2 / 330
Page 3 / 330
Page 4 / 330
Page 5 / 330
Page 6 / 330
Page 7 / 330
Page 8 / 330
Page 9 / 330
Page 10 / 330
Page 11 / 330
Page 12 / 330
Page 13 / 330
Page 14 / 330
Page 15 / 330
Page 16 / 330
Page 17 / 330
Page 18 / 330
Page 19 / 330
Page 20 / 330
Page 21 / 330
Page 22 / 330
Page 23 / 330
Page 24 / 330
Page 25 / 330
Page 26 / 330
Page 27 / 330
Page 28 / 330
Page 29 / 330
Page 30 / 330
Page 31 / 330
Page 32 / 330
Page 33 / 330
Page 34 / 330
Page 35 / 330
Page 36 / 330
Page 37 / 330
Page 38 / 330
Page 39 / 330
Page 40 / 330
Page 41 / 330
Page 42 / 330
Page 43 / 330
Page 44 / 330
Page 45 / 330
Page 46 / 330
Page 47 / 330
Page 48 / 330
Page 49 / 330
Page 50 / 330
Page 51 / 330
Page 52 / 330
Page 53 / 330
Page 54 / 330
Page 55 / 330
Page 56 / 330
Page 57 / 330
Page 58 / 330
Page 59 / 330
Page 60 / 330
Page 61 / 330
Page 62 / 330
Page 63 / 330
Page 64 / 330
Page 65 / 330
Page 66 / 330
Page 67 / 330
Page 68 / 330
Page 69 / 330
Page 70 / 330
Page 71 / 330
Pa

In [32]:
# The number of rows is the same as the number of entries reported in the API
print(sanitation_df.shape[0] == total_entries)
sanitation_df

True


countryiso3code  date      value unit obs_status  decimal    indicator.id   
0              AFE  2021        NaN                        0  SH.STA.BASS.ZS  \
1              AFE  2020  30.835943                        0  SH.STA.BASS.ZS   
2              AFE  2019  30.655874                        0  SH.STA.BASS.ZS   
3              AFE  2018   30.29641                        0  SH.STA.BASS.ZS   
4              AFE  2017  30.018284                        0  SH.STA.BASS.ZS   
..             ...   ...        ...  ...        ...      ...             ...   
37             ZWE  1964        NaN                        0  SH.STA.BASS.ZS   
38             ZWE  1963        NaN                        0  SH.STA.BASS.ZS   
39             ZWE  1962        NaN                        0  SH.STA.BASS.ZS   
40             ZWE  1961        NaN                        0  SH.STA.BASS.ZS   
41             ZWE  1960        NaN                        0  SH.STA.BASS.ZS   

                                      indicator.value country.id   
0   People using at least basic sanitation service...         ZH  \
1   People using at least basic sanitation service...         ZH   
2   People using at least basic sanitation service...         ZH   
3   People using at least basic sanitation service...         ZH   
4   People using at least basic sanitation service...         ZH   
..                                                ...        ...   
37  People using at least basic sanitation service...         ZW   
38  People using at least basic sanitation service...         ZW   
39  People using at least basic sanitation service...         ZW   
40  People using at least basic sanitation service...         ZW   
41  People using at least basic sanitation service...         ZW   

                  country.value  page  
0   Africa Eastern and Southern     0  
1   Africa Eastern and Southern     0  
2   Africa Eastern and Southern     0  
3   Africa Eastern and Southern     0  
4   Africa Eastern and Southern     0  
..                          ...   ...  
37                     Zimbabwe   329  
38                     Zimbabwe   329  
39                     Zimbabwe   329  
40                     Zimbabwe   329  
41                     Zimbabwe   329  

[16492 rows x 11 columns]

In [29]:
# Exporting dataset 
sanitation_df.to_pickle("sanitation_df.pkl")  